src/compiler/checker.ts

```diff
@@ -20131,6 +20131,11 @@ namespace ts {
                 case SyntaxKind.ParenthesizedExpression:
                 case SyntaxKind.NonNullExpression:
                     return isMatchingReference(source, (target as NonNullExpression | ParenthesizedExpression).expression);
+                case SyntaxKind.BinaryExpression:
+                    if (isAssignmentExpression(target)) {

```

Comment:
Couldn't this just be

```ts
return isAssignmentExpression(target) && isMatchingReference(source, target.left);
```

Reply:
Not all binary expressions that are narrowing expressions are assignment expressions. Being that exclusive might break cases using `instanceof`, `in`, or `,`.

Reply:
After reviewing `isMatchingReference` further, it doesn't seem like those other binary expressions can be narrowable references, so I'll go ahead with this change.